File: 2025-10-October-GFG-POTD/October 10 - ZigZag Tree Traversal/Solution.py
# Definition for a binary tree node.
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class Solution:
    def zigZagTraversal(self, root):
        if not root:
            return []
        
        from collections import deque
        result = []
        queue = deque([root])
        left_to_right = True
        
        while queue:
            size = len(queue)
            level = []
            
            for _ in range(size):
                node = queue.popleft()
                level.append(node.data)
                
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            
            if not left_to_right:
                level.reverse()
            
            result.extend(level)
            left_to_right = not left_to_right
        
        return result

File: 2025-10-October-GFG-POTD/October 10 - ZigZag Tree Traversal/Solution.cpp
/*
struct Node {
    int data;
    Node* left;
    Node* right;
    Node(int val) {
        data = val;
        left = right = nullptr;
    }
};
*/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> zigZagTraversal(Node* root) {
        vector<int> result;
        if (!root) return result;
        
        queue<Node*> q;
        q.push(root);
        bool leftToRight = true;
        
        while (!q.empty()) {
            int size = q.size();
            vector<int> level;
            
            for (int i = 0; i < size; i++) {
                Node* curr = q.front();
                q.pop();
                level.push_back(curr->data);
                
                if (curr->left) q.push(curr->left);
                if (curr->right) q.push(curr->right);
            }
            
            if (!leftToRight)
                reverse(level.begin(), level.end());
            
            result.insert(result.end(), level.begin(), level.end());
            leftToRight = !leftToRight; // toggle direction
        }
        
        return result;
    }
};
